batch.py

```diff
@@ -0,0 +1,324 @@
+    - Path to plink prefix for genotype files (plink format: .bed, .bim, .fam)
+    - Path to hail table with variant (rows) stats for downstream analyses
+    """
+    # read hail matrix table object (pre-filtered)
+    init_batch()
+    mt = hl.read_matrix_table(mt_path)
+
+    gene_name = gene_details["gene_name"]
+
+    # get relevant chromosome
+    chrom = gene_details["chr"]
+
+    # subset to window
+    # get gene body position (start and end) and build interval
+    left_boundary = max(1, int(gene_details["start"]) - window_size)
+    right_boundary = min(
+        int(gene_details["end"]) + window_size,
+        hl.get_reference("GRCh38").lengths[f"chr{chrom}"],
+    )
+    # get gene-specific genomic interval
+    gene_interval = f"{chrom}:{left_boundary}-{right_boundary}"
+    logging.info(f"Interval considered: {gene_interval}")  # 'chr22:23219960-23348287'
+
+    # include variants up to {window size} up- and downstream
+    mt = hl.filter_intervals(
+        mt, [hl.parse_locus_interval(gene_interval, reference_genome="GRCh38")]
+    )
+    mt_path = output_path(f"{gene_name}_in_window.mt", "tmp")
+    mt = mt.checkpoint(
+        mt_path, overwrite=True
+    )  # add checkpoint to avoid repeat evaluation
+    logging.info(f"Number of variants within interval: {mt.count()[0]}")
+
+    # annotate using VEP
+    vep_ht = hl.read_table(ht_path)
+    mt = mt.annotate_rows(vep=vep_ht[mt.row_key].vep)
+
+    # filter variants found to be in promoter regions
+    mt = mt.filter_rows(
+        mt.vep.regulatory_feature_consequences["biotype"].contains("promoter")
+    )
+    promoter_path = output_path(f"{gene_name}promoter_variants.mt", "tmp")
+    mt = mt.checkpoint(promoter_path, overwrite=True)  # checkpoint
+    logging.info(
+        f"Number of rare (freq<5%) QC-passing, biallelic SNPs in promoter regions: {mt.count()[0]}"
+    )
+
+    # export this as a Hail table for downstream analysis
+    ht_path = output_path(f"{gene_name}_rare_promoter_summary.ht")
+    ht = mt.rows()
+    ht.write(ht_path)
+
+    # export MT object to PLINK (promoter variants)
+    plink_path = output_path(f"{plink_output_prefix}/{gene_name}")
+    export_plink(mt, plink_path, ind_id=mt.s)
+
+    return plink_path
+
+
+# endregion GET_GENE_SPECIFIC_VARIANTS
+
+# region MISCELLANEOUS
+
+
+def make_gene_loc_dict(file) -> dict[str, dict]:
+    """
+    Turn gene information into a dictionary
+    to avoid opening this file for every gene
+    """
+    gene_dict = {}
+
+    with open(file) as handle:
+        reader = DictReader(handle, delimiter="\t")
+
+        for row in reader:
+            gene_dict[row["gene_name"]] = row
+
+    return gene_dict
+
+
+# copied from https://github.com/populationgenomics/tob-wgs/blob/main/scripts/eqtl_hail_batch/launch_eqtl_spearman.py
+# check whether it needs modifying
+def remove_sc_outliers(df):
+    """
+    Remove outlier samples, as identified by sc analysis
+    """
+    outliers = ["966_967", "88_88"]
```

Comment:
Should this be a parameter instead of hardcoding these values?

Reply:
yes good point